File: services/sns/src/models/post.model.js
const { query, getClient } = require('../config/database');

class Post {
  /**
   * Create a new post
   * @param {Number} userId - Author user ID
   * @param {Object} postData - Post data
   * @returns {Object} Created post
   */
  static async create(userId, { content, category, tags, visibility, imageUrls }) {
    const client = await getClient();
    try {
      await client.query('BEGIN');

      const sql = `
        INSERT INTO sns_posts (user_id, content, category, tags, visibility, image_urls, created_at)
        VALUES ($1, $2, $3, $4, $5, $6, NOW())
        RETURNING *
      `;

      const values = [
        userId,
        content,
        category || 'general',
        tags || [],
        visibility || 'public',
        imageUrls || []
      ];

      const result = await client.query(sql, values);

      // Increment user post_count
      await client.query(
        'UPDATE users SET post_count = COALESCE(post_count, 0) + 1 WHERE id = $1',
        [userId]
      );

      await client.query('COMMIT');

      console.log(`[SNS] Post created: ${result.rows[0].id} by user ${userId}`);
      return result.rows[0];
    } catch (error) {
      await client.query('ROLLBACK');
      console.error('[SNS] Create post error:', error);
      throw error;
    } finally {
      client.release();
    }
  }

  /**
   * Find post by ID with author info
   * @param {Number} id - Post ID
   * @returns {Object|null} Post object or null
   */
  static async findById(id) {
    const sql = `
      SELECT p.*,
             u.name AS author_name,
             u.profile_image_url AS author_image
      FROM sns_posts p
      JOIN users u ON p.user_id = u.id
      WHERE p.id = $1 AND p.is_deleted = false
    `;

    const result = await query(sql, [id]);
    return result.rows[0] || null;
  }

  /**
   * Get feed - posts from followed users
   * Cursor-based pagination using created_at
   * @param {Number} userId - Current user ID
   * @param {Object} options - Pagination options
   * @returns {Array} Posts array
   */
  static async getFeed(userId, { cursor, limit = 20 }) {
    const params = [userId, userId, limit];
    let cursorClause = '';

    if (cursor) {
      cursorClause = 'AND p.created_at < $4';
      params.push(cursor);
    }

    const sql = `
      SELECT p.*,
             u.name AS author_name,
             u.profile_image_url AS author_image,
             EXISTS(SELECT 1 FROM post_likes WHERE post_id = p.id AND user_id = $1) AS is_liked
      FROM sns_posts p
      JOIN users u ON p.user_id = u.id
      WHERE p.is_deleted = false
        AND (
          p.user_id IN (SELECT following_id FROM user_follows WHERE follower_id = $1)
          OR p.user_id = $1
        )
        AND p.user_id NOT IN (SELECT blocked_id FROM user_blocks WHERE blocker_id = $2)
        ${cursorClause}
      ORDER BY p.created_at DESC
      LIMIT $3
    `;

    const result = await query(sql, params);
    return result.rows;
  }

  /**
   * Get discover - public posts for discovery
   * Excludes blocked users
   * @param {Object} options - Filter and pagination options
   * @returns {Array} Posts array
   */
  static async getDiscover({ cursor, limit = 20, category, userId }) {
    const params = [limit];
    let paramIndex = 2;
    let cursorClause = '';
    let categoryClause = '';
    let blockClause = '';
    let likedClause = 'false AS is_liked';

    if (userId) {
      blockClause = `AND p.user_id NOT IN (SELECT blocked_id FROM user_blocks WHERE blocker_id = $${paramIndex})`;
      params.push(userId);
      paramIndex++;

      likedClause = `EXISTS(SELECT 1 FROM post_likes WHERE post_id = p.id AND user_id = $${paramIndex}) AS is_liked`;
      params.push(userId);
      paramIndex++;
    }

    if (category) {
      categoryClause = `AND p.category = $${paramIndex}`;
      params.push(category);
      paramIndex++;
    }

    if (cursor) {
      cursorClause = `AND p.created_at < $${paramIndex}`;
      params.push(cursor);
      paramIndex++;
    }

    const sql = `
      SELECT p.*,
             u.name AS author_name,
             u.profile_image_url AS author_image,
             ${likedClause}
      FROM sns_posts p
      JOIN users u ON p.user_id = u.id
      WHERE p.is_deleted = false
        AND p.visibility = 'public'
        ${blockClause}
        ${categoryClause}
        ${cursorClause}
      ORDER BY p.created_at DESC
      LIMIT $1
    `;

    const result = await query(sql, params);
    return result.rows;
  }

  /**
   * Get posts by a specific user
   * Respects visibility if requester is not following
   * @param {Number} targetUserId - Target user ID
   * @param {Object} options - Pagination options
   * @returns {Array} Posts array
   */
  static async getByUser(targetUserId, { cursor, limit = 20, requesterId }) {
    const params = [targetUserId, limit];
    let paramIndex = 3;
    let cursorClause = '';
    let visibilityClause = "AND p.visibility = 'public'";
    let likedClause = 'false AS is_liked';

    if (requesterId) {
      // If requester is the owner, show all posts
      if (requesterId === targetUserId) {
        visibilityClause = '';
      } else {
        // Check if requester follows the target user - show followers_only posts too
        visibilityClause = `AND (
          p.visibility = 'public'
          OR (p.visibility = 'followers' AND EXISTS(
            SELECT 1 FROM user_follows WHERE follower_id = $${paramIndex} AND following_id = $1
          ))
        )`;
        params.push(requesterId);
        paramIndex++;
      }

      likedClause = `EXISTS(SELECT 1 FROM post_likes WHERE post_id = p.id AND user_id = $${paramIndex}) AS is_liked`;
      params.push(requesterId === targetUserId ? targetUserId : requesterId);
      paramIndex++;
    }

    if (cursor) {
      cursorClause = `AND p.created_at < $${paramIndex}`;
      params.push(cursor);
      paramIndex++;
    }

    const sql = `
      SELECT p.*,
             u.name AS author_name,
             u.profile_image_url AS author_image,
             ${likedClause}
      FROM sns_posts p
      JOIN users u ON p.user_id = u.id
      WHERE p.user_id = $1
        AND p.is_deleted = false
        ${visibilityClause}
        ${cursorClause}
      ORDER BY p.created_at DESC
      LIMIT $2
    `;

    const result = await query(sql, params);
    return result.rows;
  }

  /**
   * Soft delete a post
   * @param {Number} postId - Post ID
   * @param {Number} userId - User ID (must be author)
   * @returns {Object|null} Deleted post or null
   */
  static async delete(postId, userId) {
    const client = await getClient();
    try {
      await client.query('BEGIN');

      const sql = `
        UPDATE sns_posts
        SET is_deleted = true, updated_at = NOW()
        WHERE id = $1 AND user_id = $2 AND is_deleted = false
        RETURNING *
      `;

      const result = await client.query(sql, [postId, userId]);

      if (result.rows[0]) {
        // Decrement user post_count
        await client.query(
          'UPDATE users SET post_count = GREATEST(COALESCE(post_count, 0) - 1, 0) WHERE id = $1',
          [userId]
        );
      }

      await client.query('COMMIT');

      return result.rows[0] || null;
    } catch (error) {
      await client.query('ROLLBACK');
      console.error('[SNS] Delete post error:', error);
      throw error;
    } finally {
      client.release();
    }
  }

  /**
   * Increment like count
   * @param {Number} postId - Post ID
   */
  static async incrementLikeCount(postId) {
    const sql = `
      UPDATE sns_posts
      SET like_count = COALESCE(like_count, 0) + 1
      WHERE id = $1
    `;
    await query(sql, [postId]);
  }

  /**
   * Decrement like count
   * @param {Number} postId - Post ID
   */
  static async decrementLikeCount(postId) {
    const sql = `
      UPDATE sns_posts
      SET like_count = GREATEST(COALESCE(like_count, 0) - 1, 0)
      WHERE id = $1
    `;
    await query(sql, [postId]);
  }

  /**
   * Increment comment count
   * @param {Number} postId - Post ID
   */
  static async incrementCommentCount(postId) {
    const sql = `
      UPDATE sns_posts
      SET comment_count = COALESCE(comment_count, 0) + 1
      WHERE id = $1
    `;
    await query(sql, [postId]);
  }

  /**
   * Decrement comment count
   * @param {Number} postId - Post ID
   */
  static async decrementCommentCount(postId) {
    const sql = `
      UPDATE sns_posts
      SET comment_count = GREATEST(COALESCE(comment_count, 0) - 1, 0)
      WHERE id = $1
    `;
    await query(sql, [postId]);
  }
}

module.exports = Post;

File: mobile/lemon_korean/lib/data/models/post_model.dart
/// Post Author model
/// Represents the author of a post or comment
class PostAuthor {
  final int id;
  final String name;
  final String? profileImageUrl;

  PostAuthor({
    required this.id,
    required this.name,
    this.profileImageUrl,
  });

  factory PostAuthor.fromJson(Map<String, dynamic> json) {
    return PostAuthor(
      id: json['id'] ?? json['user_id'] ?? 0,
      name: json['name'] ?? json['author_name'] ?? '',
      profileImageUrl: json['profile_image_url'] ?? json['author_image'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'profile_image_url': profileImageUrl,
    };
  }

  @override
  String toString() => 'PostAuthor(id: $id, name: $name)';

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is PostAuthor && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;
}

/// Post model
/// Represents a social post in the SNS feed
class PostModel {
  final int id;
  final PostAuthor author;
  final String content;
  final String category; // 'learning' or 'general'
  final List<String> tags;
  final String visibility; // 'public' or 'followers'
  final List<String> imageUrls;
  final int likeCount;
  final int commentCount;
  final bool isLiked;
  final DateTime createdAt;

  PostModel({
    required this.id,
    required this.author,
    required this.content,
    required this.category,
    this.tags = const [],
    this.visibility = 'public',
    this.imageUrls = const [],
    this.likeCount = 0,
    this.commentCount = 0,
    this.isLiked = false,
    required this.createdAt,
  });

  factory PostModel.fromJson(Map<String, dynamic> json) {
    return PostModel(
      id: json['id'] as int,
      author: PostAuthor.fromJson(json),
      content: (json['content'] ?? '') as String,
      category: (json['category'] ?? 'general') as String,
      tags: json['tags'] != null
          ? List<String>.from(json['tags'] as List)
          : [],
      visibility: (json['visibility'] ?? 'public') as String,
      imageUrls: json['image_urls'] != null
          ? List<String>.from(json['image_urls'] as List)
          : [],
      likeCount: (json['like_count'] ?? 0) as int,
      commentCount: (json['comment_count'] ?? 0) as int,
      isLiked: (json['is_liked'] ?? false) as bool,
      createdAt: json['created_at'] != null
          ? DateTime.parse(json['created_at'] as String)
          : DateTime.now(),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user_id': author.id,
      'name': author.name,
      'profile_image_url': author.profileImageUrl,
      'content': content,
      'category': category,
      'tags': tags,
      'visibility': visibility,
      'image_urls': imageUrls,
      'like_count': likeCount,
      'comment_count': commentCount,
      'is_liked': isLiked,
      'created_at': createdAt.toIso8601String(),
    };
  }

  PostModel copyWith({
    int? id,
    PostAuthor? author,
    String? content,
    String? category,
    List<String>? tags,
    String? visibility,
    List<String>? imageUrls,
    int? likeCount,
    int? commentCount,
    bool? isLiked,
    DateTime? createdAt,
  }) {
    return PostModel(
      id: id ?? this.id,
      author: author ?? this.author,
      content: content ?? this.content,
      category: category ?? this.category,
      tags: tags ?? this.tags,
      visibility: visibility ?? this.visibility,
      imageUrls: imageUrls ?? this.imageUrls,
      likeCount: likeCount ?? this.likeCount,
      commentCount: commentCount ?? this.commentCount,
      isLiked: isLiked ?? this.isLiked,
      createdAt: createdAt ?? this.createdAt,
    );
  }

  // Helpers
  bool get isLearningPost => category == 'learning';
  bool get isGeneralPost => category == 'general';
  bool get isPublic => visibility == 'public';
  bool get hasImages => imageUrls.isNotEmpty;
  bool get hasTags => tags.isNotEmpty;

  @override
  String toString() => 'PostModel(id: $id, author: ${author.name}, content: ${content.length > 30 ? '${content.substring(0, 30)}...' : content})';

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is PostModel && other.id == id;
  }

  @override
  int get hashCode => id.hashCode;
}
